src/webrtc/opcodes/Video.ts

```diff
@@ -15,137 +15,109 @@
-function handleSSRC(this: WebSocket, type: "audio" | "video", ssrcs: SSRCs) {
-	if (!this.client) return;
-	const stream = this.client.in.stream!;
-	const transport = this.client.transport!;
+	const ssrcs = this.webRtcClient.getIncomingStreamSSRCs();
 
-	const id = type + ssrcs.media;
-	var track = stream.getTrack(id);
-	if (!track) {
-		console.log("createIncomingStreamTrack", id);
-		track = transport.createIncomingStreamTrack(type, { id, ssrcs });
-		stream.addTrack(track);
+	const clientsThatNeedUpdate = new Set<WebRtcClient<WebRtcWebSocket>>();
 
-		const clients = getClients(this.client.channel_id)!;
-		clients.forEach((client) => {
-			if (client.websocket.user_id === this.user_id) return;
-			if (!client.out.stream) return;
+	// check if client has signaled that it will send audio
+	if (d.audio_ssrc !== 0) {
+		// check if we already have incoming media for this ssrcs, if not, publish a new audio track for it
+		if (ssrcs.audio_ssrc != d.audio_ssrc) {
+			console.log(
+				`[${this.user_id}] publishing new audio track ssrc:${d.audio_ssrc}`,
+			);
+			this.webRtcClient.publishTrack("audio", {
+				audio_ssrc: d.audio_ssrc,
+			});
+		}
+
+		// now check that all clients have outgoing media for this ssrcs
+		for (const client of mediaServer.getClientsForRtcServer<WebRtcWebSocket>(
+			rtc_server_id,
+		)) {
+			if (client.user_id === this.user_id) continue;
+
+			const ssrcs = client.getOutgoingStreamSSRCsForUser(this.user_id);
+			if (ssrcs.audio_ssrc != d.audio_ssrc) {
+				console.log(
+					`[${client.user_id}] subscribing to audio track ssrcs: ${d.audio_ssrc}`,
+				);
+				client.subscribeToTrack(this.webRtcClient.user_id, "audio");
+
+				clientsThatNeedUpdate.add(client);
+			}
+		}
+	}
+	// check if client has signaled that it will send video
+	if (d.video_ssrc !== 0 && stream?.active) {
+		// check if we already have incoming media for this ssrcs, if not, publish a new video track for it
+		if (ssrcs.video_ssrc != d.video_ssrc) {
+			console.log(
+				`[${this.user_id}] publishing new video track ssrc:${d.video_ssrc}`,
+			);
+			this.webRtcClient.publishTrack("video", {
+				video_ssrc: d.video_ssrc,
+				rtx_ssrc: d.rtx_ssrc,
+			});
+		}
+
+		// now check that all clients have outgoing media for this ssrcs
+		for (const client of mediaServer.getClientsForRtcServer<WebRtcWebSocket>(
+			rtc_server_id,
+		)) {
+			if (client.user_id === this.user_id) continue;
+
+			const ssrcs = client.getOutgoingStreamSSRCsForUser(
+				this.webRtcClient.user_id,
+			);
+			if (ssrcs.video_ssrc != d.video_ssrc) {
+				console.log(
+					`[${client.user_id}] subscribing to video track ssrc: ${d.video_ssrc}`,
+				);
+				client.subscribeToTrack(this.webRtcClient.user_id, "video");
+
+				clientsThatNeedUpdate.add(client);
+			}
+		}
+	}
 
-			attachTrack.call(this, track, client.websocket.user_id);
+	for (const client of clientsThatNeedUpdate) {
+		const ssrcs = client.getOutgoingStreamSSRCsForUser(this.user_id);
+
+		Send(client.websocket, {
```

Comment:
should await this so that rejects can be captured. perhaps Promise.all them after this for loop?

Reply:
Updated to use Promise.all